elastic/elastic-agent
# [Upgrade Watcher][Crash Checker] Consider Agent process as crashed if its PID remains 0

## What does this PR do?

This PR fixes a bug in the Upgrade Watcher's Crash Checker where it was consider the Agent process as healthy (not crashed) despite its PID remaining 0 every time the Crash Checker retrieved the PID from the service.

## Why is it important?

To detect when Agent has crashed so the Upgrade Watcher can initiate a rollback.

## Checklist




## How to test this PR locally

_Manually testing this PR is currently blocked on https://github.com/elastic/elastic-agent/issues/3377._

Testing this PR locally is not trivial, but it's possible.  It involves upgrading to an Agent build where the Agent binary deliberately crashes, and making sure that the Agent is then rolled back to the previous (pre-upgrade) version.

1. Install and run Elastic Agent >= `8.10.0`. This is necessary because we need the pre-upgrade Agent to kick off the Upgrade Watcher using the post-upgrade Agent's binary, a change that was [implemented in `8.10.0`](https://github.com/elastic/elastic-agent/pull/3140).

2. Checkout this PR and make changes to the `elastic-agent run` code path such that the Agent process will exit with an error.  The easiest change would probably be to add a small sleep, say 5 seconds, followed by returning an error, right here: https://github.com/elastic/elastic-agent/blob/539a5f2956e423558b2d48fdb8bcdcf257348d51/internal/pkg/agent/cmd/run.go#L65-L66

3. Also bump up the Agent version to `8.12.0` over here so the upgrade is possible: https://github.com/elastic/elastic-agent/blob/243c76b3ebec56130bd6bd172f5d6725d289fbca/version/version.go#L7

4. Build the Agent package with the changes. Since the Agent version has been bumped up and the corresponding component binaries won't be available, make sure to set `AGENT_DROP_PATH` to nothing. Make sure NOT to use `SNAPSHOT=true` otherwise the snapshot artifact downloader will kick in during the upgrade process.
   ```
   DEV=true AGENT_DROP_PATH= PLATFORMS=linux/arm64 PACKAGES=tar.gz mage package
   ```
5. Upgrade the running Agent to the built Agent.
   ```
   sudo elastic-agent upgrade 8.12.0 --source-uri file:///home/shaunak/development/github/elastic-agent/build/distributions/ --skip-verify
   ```
6. Ensure that the Agent is upgrading.
   ```
   sudo elastic-agent status
   ```
7. While the upgrade is in progress, watch the Upgrade Watcher's log.
8. After a couple of minutes, check that the Agent was rolled back.
   ```
   sudo elastic-agent version
   ```

## Related issues

- Closes #3124
